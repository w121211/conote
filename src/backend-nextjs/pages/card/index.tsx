import { useRouter } from 'next/router'
import { symbolToUrl, getCardUrlParam } from '../../lib/helper'
import { CocardQuery, useCocardQuery } from '../../apollo/query.graphql'
import { QueryDataProvider } from '../../components/data-provider'
<<<<<<< HEAD
import { Button } from 'antd'
import TickerAnchor from '../../components/anchor/tickerAnchor'
import { CardBody, CardHead } from '../../components/card'
import Discuss from '../../components/discuss/discuss'
import useMe from '../../components/use-me'
import EditIcon from '../../assets/svg/edit.svg'
import classes from './card-page.module.scss'
import { useState } from 'react'
=======
import { CardBody, CardHead } from '../../components/card'
import useMe from '../../components/use-me'
>>>>>>> backend-dev

function CardPage(): JSX.Element {
  const me = useMe({ redirectTo: '/signin' })
  const router = useRouter()
  const { u, s } = router.query
  const url = u as string
  const symbol = s as string
<<<<<<< HEAD
  // const [Question, setQuestion] = useState(false)
  const [discuss, setDiscuss] = useState(true)

  const discussClickLHandler = () => {
    setDiscuss(true)
  }
  const discussClickRHandler = () => {
    setDiscuss(false)
  }

  let titleRefArr: any[] = []
  const clickHandler = (e: any, index: number) => {
    titleRefArr[index].scrollIntoView({ behavior: 'smooth' }, 500)
  }
  const titleRefHandler = (arr: any[]) => {
    titleRefArr = arr
  }
  const showQuestion = () => {
    setDiscuss(false)
  }
=======
>>>>>>> backend-dev

  function _render(url: string): JSX.Element {
    return (
      <QueryDataProvider
        useQuery={() => useCocardQuery({ variables: { url } })}
        render={(data: CocardQuery) => {
          if (data && data.cocard) {
            const url = `/card/form?${getCardUrlParam(data.cocard)}`
            return (
<<<<<<< HEAD
              <div className={classes.main}>
                <div className={classes.mainInner}>
                  <TickerAnchor data={data.cocard} clickHandler={clickHandler} />
                  <div className={classes.cardOuter}>
                    <div className={classes.cardInner}>
                      <div className={classes.cardElement}>
                        <CardHead card={data.cocard} />
                        <CardBody card={data.cocard} titleRefHandler={titleRefHandler} showQuestion={showQuestion} />
                      </div>
                    </div>
                  </div>
                  <Button
                    className={classes.button}
                    onClick={() => {
                      router.push(url)
                    }}
                  >
                    {console.log(url)}
                    <EditIcon className={classes.editIcon} />
                  </Button>
                  {/* <button
                    onClick={() => {
                      router.push(url)
                    }}
                  >
                    編輯
                  </button> */}
                </div>
                <Discuss
                  switchTab={discuss}
                  discussClickLHandler={discussClickLHandler}
                  discussClickRHandler={discussClickRHandler}
                />
=======
              <div>
                <CardHead card={data.cocard} />
                <CardBody card={data.cocard} />
                <button
                  onClick={() => {
                    router.push(url)
                  }}
                >
                  編輯
                </button>
>>>>>>> backend-dev
              </div>
            )
          }
          return null
        }}
      />
    )
  }

  if (url) {
    return _render(url)
  }
  if (symbol) {
    try {
      return _render(symbolToUrl(symbol))
    } catch {
      return <h1>Symbol format error</h1>
    }
  }
  return <h1>Require URL or Symbol</h1>
}

export default CardPage
