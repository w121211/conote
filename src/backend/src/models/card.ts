// import * as _ from 'lodash';
// import dayjs from 'dayjs';
import * as PA from '@prisma/client'
import { prisma } from '../context'
import { getBotEmail } from './user'
import { getOrCreateSymbol, symbolToUrl, SYMBOL_DOMAIN } from './symbol'
import { TextEditor, MarkToConnectedContentRecord } from '../../../lib/editor/src'

interface CardTemplate {
  body: string
  connContents: MarkToConnectedContentRecord
}

export interface CardMeta {
  symbol?: string
  // 紀錄marker-line與connected-content（poll, comment, etc)
  conn: MarkToConnectedContentRecord
  // 每個card都帶有一個comment作為discuss-board
  commentId: number
}

const TEMPLATE: Record<PA.CardTemplate, CardTemplate> = {
  [PA.CardTemplate.TICKER]: {
    body: `[=]
[?:poll] <買> vs <賣>
[+]
[-]
[VS]
- [[<$AAA> vs <$BBB>]]
- [[<$AAA> vs <$CCC>]]
[Q]
`,
    connContents: { '[?:poll]': { poll: true, pollChoices: ['買', '賣'] } },
  },
  [PA.CardTemplate.TOPIC]: {
    body: `[=]
[?:poll] <看多> vs <看空>
[+]
[-]
[Q]`,
    connContents: { '[?:poll]': { comment: true, pollChoices: ['看多', '看空'] } },
  },
  [PA.CardTemplate.WEBPAGE]: {
    body: '',
    connContents: {},
  },
}

const SYMBOL_TO_TEMPLATE: Record<PA.SymbolCat, PA.CardTemplate> = {
  [PA.SymbolCat.TICKER]: PA.CardTemplate.TICKER,
  [PA.SymbolCat.TOPIC]: PA.CardTemplate.TOPIC,
}

export async function createConnectedContents(
  contents: MarkToConnectedContentRecord,
): Promise<MarkToConnectedContentRecord> {
  const record: MarkToConnectedContentRecord = {}
  for (const k in contents) {
    const e = contents[k]
    if (e.poll && e.pollChoices) {
      // eslint-disable-next-line no-await-in-loop
      const comment = await prisma.comment.create({
        data: {
          // parse marker-value作為poll的內文及選項
          text: k,
          user: { connect: { email: getBotEmail() } },
          count: { create: {} },
          poll: {
            create: {
              choices: e.pollChoices,
              user: { connect: { email: getBotEmail() } },
              count: { create: {} },
            },
          },
        },
        include: { poll: true },
      })
      record[k] = { ...e, commentId: comment.id, pollId: comment.poll?.id }
    } else if (e.comment) {
      // eslint-disable-next-line no-await-in-loop
      const comment = await prisma.comment.create({
        data: {
          // 直接用marker作為text
          text: k,
          user: { connect: { email: getBotEmail() } },
          count: { create: {} },
        },
      })
      record[k] = { ...e, commentId: comment.id }
    }
  }
  return record
}

async function createCard(
  template: PA.CardTemplate,
  symbol?: PA.Symbol,
  link?: PA.Link,
): Promise<
  PA.Cocard & {
    link: PA.Link
    body: PA.CardBody
  }
> {
  if (symbol === undefined && link === undefined) {
    throw new Error('`symbol`或`link`需要有其中一個')
  }
  const conn = await createConnectedContents(TEMPLATE[template].connContents)

  // 創body，因為是第一次創，不使用`createCardBody()`
  // 將connected-contents加到markerlines中
  const editor = new TextEditor()
  editor.setBody(TEMPLATE[template].body)
  editor.flush()
  editor.addConnectedContents(conn)
  const body = await prisma.cardBody.create({
    data: {
      text: editor.toStoredText(),
      user: { connect: { email: getBotEmail() } },
    },
  })

  // 創card專屬comment
  const cardComment = await prisma.comment.create({
    data: {
      text: '',
      user: { connect: { email: getBotEmail() } },
      count: { create: {} },
    },
  })

  // 創meta
  const meta: CardMeta = {
    symbol: symbol ? symbol.name : undefined,
    conn,
    commentId: cardComment.id,
  }

  // 創card, link
  return await prisma.cocard.create({
    data: {
      template,
      meta: meta as Record<string, any>,
      body: { connect: { id: body.id } },
      link: symbol
        ? { create: { url: symbolToUrl(symbol.name), domain: SYMBOL_DOMAIN } }
        : { connect: { id: (link as PA.Link).id } },
    },
    include: {
      link: true,
      body: true,
    },
  })
}

export async function createCardBody(card: PA.Cocard, editor: TextEditor, userId: string): Promise<PA.CardBody> {
  const creaters: PA.Prisma.Prisma__AnchorClient<PA.Anchor>[] = []
  for (const e of editor.getMarkerlines()) {
    if (e.comment && e.commentId && e.marker?.value) {
      // TODO: 對comment創reply
      // await prisma.reply.create({})
    }
    if (e.new && e.stampId) {
      // 創anchor
      creaters.push(
        prisma.anchor.create({
          data: {
            // user: { connect: { email: userEmail } },
            user: { connect: { id: userId } },
            cocard: { connect: { id: card.id } },
            // cocard: isOcard(card) ? undefined : { connect: { id: card.id } },
            // ocard: isOcard(card) ? { connect: { id: card.id } } : undefined,
            count: { create: {} },
            path: e.stampId,
          },
        }),
      )
    }
  }
  const anchors = (await prisma.$transaction([...creaters])) as PA.Anchor[]
  editor.addAnchors(
    anchors.map(e => {
      if (e.path === null) throw new Error()
      return { ...e, stamp: e.path }
    }),
  )

  // 創新body，連結prev-body（prev-body會自動與cocard斷開連結）
  return await prisma.cardBody.create({
    data: {
      text: editor.toStoredText(),
      user: { connect: { id: userId } },
      cocard: { connect: { id: card.id } },
      prev: { connect: { id: card.bodyId } },
    },
  })
}

export async function getOrCreateCardBySymbol(
  symbolName: string,
): Promise<
  PA.Cocard & {
    link: PA.Link
    body: PA.CardBody
  }
> {
  const found = await prisma.cocard.findUnique({
    where: { linkUrl: symbolToUrl(symbolName) },
    include: { link: true, body: true },
  })
  if (found) {
    return { ...found }
    // const [text, linemetas] = splitMetatext(card.body.text);
    // return { ...card, body: { ...card.body, text, linemetas } };
  }

  // 沒找到cocard，創一個，同步創link
  const [symbol] = await getOrCreateSymbol(symbolName)
  return await createCard(SYMBOL_TO_TEMPLATE[symbol.cat], symbol)
}

export async function getOrCreateCardByLink(
  link: PA.Link,
): Promise<
  PA.Cocard & {
    link: PA.Link
    body: PA.CardBody
  }
> {
  // 找cocard
  const found = await prisma.cocard.findUnique({
    where: { linkUrl: link.url },
    include: { link: true, body: true },
  })
  if (found) {
    return found
    // const [text, linemetas] = splitMetatext(card.body.text);
    // return { ...card, body: { ...card.body, text, linemetas } };
  }

  // 沒找到cocard，創一個
  return await createCard(PA.CardTemplate.WEBPAGE, undefined, link)
}
