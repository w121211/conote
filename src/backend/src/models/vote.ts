import * as PA from '@prisma/client'
import { prisma } from '../context'

export async function createVote(poll: PA.Poll, choice: string, userId: string, oauthorName: string): Promise<PA.Vote> {
  // TODO: 1. user一段時間內不能重複vote 2. oauther同一個來源、1個poll只能有1個vote

  // Validation
  const choiceIdx = poll.choices.indexOf(choice)
  if (choiceIdx < 0) throw new Error('poll中沒有這個vote choice')

  return await prisma.vote.create({
    data: {
      choiceIdx,
      user: { connect: { id: userId } },
      oauthor: { connect: { name: oauthorName } },
      poll: { connect: { id: poll.id } },
    },
  })
}
