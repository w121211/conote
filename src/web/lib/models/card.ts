import { Card, Link, Sym, SymType } from '.prisma/client'
import { FetchClient } from '../fetcher/fetcher'
import prisma from '../prisma'
import { CardStateModel, CardStateParsed } from './card-state'
import { LinkService } from './link'
import { SymModel } from './sym'

export type CardMeta = {
  template?: 'webpage' | 'ticker' | 'topic' | 'vs'
  redirects?: string[]
  duplicates?: string[]
  url?: string
  author?: string
  publishedAt?: string
  description?: string
  keywords?: string[]
  // lang?: string
  title?: string
}

export const CardModel = {
  async get(id: string): Promise<
    | null
    | (Omit<Card, 'meta'> & {
        link: Link | null
        meta: CardMeta
        state: CardStateParsed | null
        sym: Sym
      })
  > {
    const card = await prisma.card.findUnique({
      where: { id },
      include: {
        link: true,
        sym: true,
      },
    })
    if (card) {
      return {
        ...card,
        meta: card.meta as unknown as CardMeta,
        state: await CardStateModel.getLastCardState(card.id),
      }
    }
    return null
  },

  /**
   * Get card by symbol
   *
   * @throw symbol parse error
   */
  async getBySymbol(symbol: string): Promise<
    | (Omit<Card, 'meta'> & {
        link: Link | null
        meta: CardMeta
        state: CardStateParsed
        sym: Sym
      })
    | null
  > {
    const parsed = SymModel.parse(symbol)
    const sym = await prisma.sym.findUnique({
      where: { symbol: parsed.symbol },
      include: { card: { include: { link: true } } },
    })
    if (sym?.card) {
      const { card, ...restSym } = sym
      const state = await CardStateModel.getLastCardState(card.id)
      if (state === null) {
        throw '[conote-web] symbol-card state cannot be null'
      }
      return {
        ...card,
        meta: card.meta as unknown as CardMeta,
        state,
        sym: restSym,
      }
    }
    return null
  },

  /**
   * Get or create a webpage-card by URL, if the URL not found in database, will try to scrape and then store
   * Newly created webpage-card will not have card-state and remains null.
   */
  async getOrCreateByUrl({ scraper, url }: { scraper?: FetchClient; url: string }): Promise<
    Omit<Card, 'meta'> & {
      link: Link
      meta: CardMeta
      state: CardStateParsed | null
      sym: Sym
    }
  > {
    const [link, { fetchResult }] = await LinkService.getOrCreateLink({ scraper, url })
    if (link.card) {
      const { card, ...linkRest } = link
      return {
        ...card,
        link: linkRest,
        meta: card.meta as unknown as CardMeta,
        state: await CardStateModel.getLastCardState(card.id),
      }
    }

    // Card not found, create one
    const meta: CardMeta = {
      template: 'webpage',
      url: link.url,
      author: fetchResult?.authorName,
      publishedAt: fetchResult?.date,
      description: fetchResult?.date,
      keywords: fetchResult?.keywords,
      title: fetchResult?.title,
    }
    const card = await prisma.card.create({
      data: {
        link: { connect: { id: link.id } },
        meta,
        sym: {
          create: {
            type: SymType.URL,
            symbol: LinkService.toSymbol(link),
          },
        },
      },
      include: { sym: true },
    })
    return {
      ...card,
      link,
      meta,
      state: null,
    }
  },
}
