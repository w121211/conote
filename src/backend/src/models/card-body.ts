/* eslint-disable no-await-in-loop */
// import * as _ from 'lodash';
// import dayjs from 'dayjs';
import * as PA from '@prisma/client'
import { Editor, DBLinker, Markerline } from '../../../lib/editor/src'
import { prisma } from '../context'
import { getBotId } from './user'
import { getOrCreateCardBySymbol } from './card'
import { CommentMeta } from './comment'

export type CardBodyMeta = Markerline[]

const NEAT_REPLY_CHOICE = [
  { options: ['<BUY>', '<B>', '<買>', '<LONG>', '<L>', '<多>', '<看多>'], choiceIdx: 0 },
  { options: ['<SELL>', '<S>', '<買>', '<SHORT>', '<空>', '<看空>'], choiceIdx: 1 },
]

async function createNeatReply(mkln: Markerline, userId: string): Promise<PA.Reply> {
  // 適用條件：在webcard裡幫oauther針對某個symbol的快速投票 -> 必須有投票
  if (!mkln.neatReply) throw new Error('非neatReply')
  if (!mkln.src) throw new Error('缺src')
  if (!mkln.stampId) throw new Error('缺stampId')
  if (!mkln.pollChoices) throw new Error('缺pollChoices')
  if (mkln.pollChoices.length !== 1) throw new Error('pollChoices的length不等於1')
  if (!mkln.nestedCard) throw new Error('缺nestedCard')
  if (!mkln.oauthor) throw new Error('缺oauthor')

  // neat-reply的choice需是預設的幾個
  const choice = mkln.pollChoices[0]
  let choiceIdx
  for (const e of NEAT_REPLY_CHOICE) {
    if (e.options.indexOf(choice)) choiceIdx = e.choiceIdx
  }
  if (choiceIdx === undefined) {
    throw new Error('所給的vote-choice非預設的那幾個')
  }

  // 取得卡片，找到對應的comment-id, poll-id
  const card = await getOrCreateCardBySymbol(mkln.nestedCard.symbol)
  const editor = new Editor(card.body.text, (card.body.meta as unknown) as Markerline[])
  const pollMkln = editor
    .getMarkerlines()
    .find(async e => e.pollId && e.commentId && e.createrId === (await getBotId()))
  if (pollMkln === undefined || pollMkln.commentId === undefined) {
    throw new Error('卡片中找不到預設poll')
  }

  // 創reply
  return await prisma.reply.create({
    data: {
      text: `${mkln.str} ^[[${mkln.src}:${mkln.stampId}]]`,
      user: { connect: { id: userId } },
      oauthor: { connect: { name: mkln.oauthor } },
      count: { create: {} },
      comment: { connect: { id: pollMkln.commentId } },
    },
  })

  // TODO: 創vote
  // prisma.vote.create()
}

export async function createCardBody(
  card: PA.Cocard,
  editor: Editor,
  userId: string,
  bodyId?: number,
): Promise<PA.CardBody> {
  const dict: Record<string, DBLinker> = {} // { [marker]: DBLinker }

  for (const e of editor.getMarkerlines()) {
    // TODO: 對於編輯過的marker也需要處理（？）
    if (e.new && e.stampId) {
      // 若是新的markerline，創anchor, comment等等

      // 創anchor
      const anchor = await prisma.anchor.create({
        data: {
          // user: { connect: { email: userEmail } },
          user: { connect: { id: userId } },
          cocard: { connect: { id: card.id } },
          // cocard: isOcard(card) ? undefined : { connect: { id: card.id } },
          // ocard: isOcard(card) ? { connect: { id: card.id } } : undefined,
          count: { create: {} },
          path: e.stampId,
        },
      })

      // 創comment, poll, reply, shortcut-reply
      let commentId, pollId, replyId
      if (e.poll && e.pollChoices && e.marker?.value) {
        const comment = await prisma.comment.create({
          data: {
            meta: ({ inCardIds: [card.id] } as CommentMeta) as any,
            text: e.marker?.value,
            user: { connect: { id: userId } },
            count: { create: {} },
            poll: {
              create: {
                choices: e.pollChoices,
                user: { connect: { id: userId } },
                count: { create: {} },
              },
            },
          },
          include: { poll: true },
        })
        commentId = comment.id
        pollId = comment.poll?.id
      } else if (e.comment && e.marker?.value) {
        const comment = await prisma.comment.create({
          data: {
            meta: ({ inCardIds: [card.id] } as CommentMeta) as any,
            text: e.marker?.value,
            user: { connect: { id: userId } },
            count: { create: {} },
          },
        })
        commentId = comment.id
      } else if (e.neatReply) {
        try {
          const reply = await createNeatReply(e, userId)
          replyId = reply.id
        } catch (err) {
          console.error(err)
        }
      }

      dict[e.stampId] = { createrId: userId, anchorId: anchor.id, commentId, pollId, replyId }
    }
  }

  // （重要）將databse所創的資訊與markerlines同步
  editor.attachDblinker(dict)

  // 更新或創新body，連結prev-body，prev-body會自動與cocard斷開連結
  if (bodyId) {
    return await prisma.cardBody.update({
      data: {
        meta: (editor.getMarkerlines() as CardBodyMeta) as any,
        text: editor.getText(),
      },
      where: { id: bodyId },
    })
  } else {
    return await prisma.cardBody.create({
      data: {
        meta: (editor.getMarkerlines() as CardBodyMeta) as any,
        text: editor.getText(),
        user: { connect: { id: userId } },
        cocard: { connect: { id: card.id } },
        prev: { connect: { id: card.bodyId } },
      },
    })
  }
}
