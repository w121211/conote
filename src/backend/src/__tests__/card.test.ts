/**
 * @jest-environment ./src/__tests__/prisma-test-environment
 */
// import { createConnectedContents } from '../models/card'
// describe('card', () => {
//   it('use url to split text to sections', () => {
//     const a =
//       'https://www.youtube.com/watch?v=qSYGlOZNUCw\n$AAAA\n\nhttps://www.youtube.com/watch?v=kprnW5VadnY&list=WL&index=53\n$BBBB\n'
//     expect(createConnectedContents().toEqual())
//   })
// })
import { createTestContext } from './__helpers'
import { createCardBody } from '../models/card'

const ctx = createTestContext()

it('ensures that a draft can be created and published', async () => {
  // Create a new draft
  const draftResult = await ctx.client.request(`               # 1
    mutation {
      createDraft(title: "Nexus", body: "...") {            # 2
        id
        title
        body
        published
      }
    }
  `)

  // Snapshot that draft and expect `published` to be false
  expect(draftResult).toMatchInlineSnapshot(`
    Object {
      "createDraft": Object {
        "body": "...",
        "id": 1,
        "published": false,
        "title": "Nexus",
      },
    }
  `) // 3

  // Publish the previously created draft
  const publishResult = await ctx.client.request(
    `
    mutation publishDraft($draftId: Int!) {
      publish(draftId: $draftId) {
        id
        title
        body
        published
      }
    }
  `,
    { draftId: draftResult.createDraft.id },
  )

  // Snapshot the published draft and expect `published` to be true
  expect(publishResult).toMatchInlineSnapshot(`
    Object {
      "publish": Object {
        "body": "...",
        "id": 1,
        "published": true,
        "title": "Nexus",
      },
    }
  `)

  const persistedData = await ctx.prisma.post.findMany()

  expect(persistedData).toMatchInlineSnapshot(`
    Array [
      Object {
        "body": "...",
        "id": 1,
        "published": true,
        "title": "Nexus",
      },
    ]
  `)
})
