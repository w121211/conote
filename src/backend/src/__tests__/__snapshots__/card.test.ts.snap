// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create a blank web card 1`] = `
Object {
  "body": Object {
    "id": 5,
    "meta": Array [],
    "prevId": 4,
    "text": "",
    "userId": "bot",
  },
  "bodyId": 5,
  "id": 2,
  "link": Object {
    "domain": "test1.com",
    "id": 2,
    "meta": Object {
      "domain": "test1.com",
      "resolvedUrl": "http://test1.com",
      "srcType": "OTHER",
    },
    "oauthorName": null,
    "srcId": null,
    "srcType": "OTHER",
    "url": "http://test1.com",
  },
  "linkUrl": "http://test1.com",
  "meta": Object {
    "commentId": 3,
  },
  "template": "WEBPAGE",
}
`;

exports[`create a nested web-card 1`] = `
"$CC
[?] <BUY> @30 %DO4
[+]
111 111 %RzE
222 222 %WGP

$DD
[?] <SELL> @30 %BBc"
`;

exports[`create a nested web-card 2`] = `
Array [
  Object {
    "anchorId": 11,
    "createrId": "test-user-2",
    "linenumber": 1,
    "marker": Object {
      "key": "[?]",
      "value": "<BUY> @30",
    },
    "neatReply": true,
    "nestedCard": Object {
      "symbol": "$CC",
    },
    "oauthor": "test-oauthor-1",
    "pollChoices": Array [
      "<BUY>",
    ],
    "replyId": 1,
    "src": "http://test2.com",
    "stampId": "%DO4",
    "str": " <BUY> @30",
  },
  Object {
    "anchorId": 12,
    "createrId": "test-user-2",
    "linenumber": 3,
    "marker": Object {
      "key": "[+]",
      "value": "111 111",
    },
    "nestedCard": Object {
      "symbol": "$CC",
    },
    "oauthor": "test-oauthor-1",
    "src": "http://test2.com",
    "stampId": "%RzE",
    "str": "111 111",
  },
  Object {
    "anchorId": 13,
    "createrId": "test-user-2",
    "linenumber": 4,
    "marker": Object {
      "key": "[+]",
      "value": "222 222",
    },
    "nestedCard": Object {
      "symbol": "$CC",
    },
    "oauthor": "test-oauthor-1",
    "src": "http://test2.com",
    "stampId": "%WGP",
    "str": "222 222",
  },
  Object {
    "anchorId": 14,
    "createrId": "test-user-2",
    "linenumber": 7,
    "marker": Object {
      "key": "[?]",
      "value": "<SELL> @30",
    },
    "neatReply": true,
    "nestedCard": Object {
      "symbol": "$DD",
    },
    "oauthor": "test-oauthor-1",
    "pollChoices": Array [
      "<SELL>",
    ],
    "replyId": 2,
    "src": "http://test2.com",
    "stampId": "%BBc",
    "str": " <SELL> @30",
  },
]
`;

exports[`create a nested web-card 3`] = `
"[~]
[?]
<BUY> <SELL> %Vre
[+]
222 222 %rcC
111 111 %PGu
[-]
"
`;

exports[`create a nested web-card 4`] = `
Array [
  Object {
    "anchorId": 7,
    "commentId": 5,
    "createrId": "bot",
    "linenumber": 2,
    "marker": Object {
      "key": "[?]",
      "value": "<BUY> <SELL>",
    },
    "poll": true,
    "pollChoices": Array [
      "<BUY>",
      "<SELL>",
    ],
    "pollId": 2,
    "stampId": "%Vre",
    "str": "<BUY> <SELL> %Vre",
  },
  Object {
    "anchorId": 8,
    "createrId": "test-user-2",
    "linenumber": 4,
    "marker": Object {
      "key": "[+]",
      "value": "222 222",
    },
    "oauthor": "test-oauthor-1",
    "src": "http://test2.com",
    "srcStamp": "%WGP",
    "stampId": "%rcC",
    "str": "222 222 %rcC",
  },
  Object {
    "anchorId": 9,
    "createrId": "test-user-2",
    "linenumber": 5,
    "marker": Object {
      "key": "[+]",
      "value": "111 111",
    },
    "oauthor": "test-oauthor-1",
    "src": "http://test2.com",
    "srcStamp": "%RzE",
    "stampId": "%PGu",
    "str": "111 111 %PGu",
  },
]
`;

exports[`create a nested web-card 5`] = `
"[~]
[?]
<BUY> <SELL> %JwE
[+]
[-]"
`;

exports[`create a nested web-card 6`] = `
Array [
  Object {
    "anchorId": 10,
    "commentId": 7,
    "createrId": "bot",
    "linenumber": 2,
    "marker": Object {
      "key": "[?]",
      "value": "<BUY> <SELL>",
    },
    "poll": true,
    "pollChoices": Array [
      "<BUY>",
      "<SELL>",
    ],
    "pollId": 3,
    "stampId": "%JwE",
    "str": "<BUY> <SELL> %JwE",
  },
]
`;

exports[`create a symbol card 1`] = `
Object {
  "body": Object {
    "id": 2,
    "meta": Array [
      Object {
        "anchorId": 1,
        "commentId": 1,
        "createrId": "bot",
        "linenumber": 2,
        "marker": Object {
          "key": "[?]",
          "value": "<BUY> <SELL>",
        },
        "poll": true,
        "pollChoices": Array [
          "<BUY>",
          "<SELL>",
        ],
        "pollId": 1,
        "stampId": "%2Md",
        "str": "<BUY> <SELL>",
      },
    ],
    "prevId": 1,
    "text": "[~]
[?]
<BUY> <SELL> %2Md
[+]
[-]",
    "userId": "bot",
  },
  "bodyId": 2,
  "id": 1,
  "link": Object {
    "domain": "_",
    "id": 1,
    "meta": null,
    "oauthorName": null,
    "srcId": null,
    "srcType": "OTHER",
    "url": "//$AA",
  },
  "linkUrl": "//$AA",
  "meta": Object {
    "commentId": 2,
    "symbol": "$AA",
  },
  "template": "TICKER",
}
`;

exports[`create a symbol card 2`] = `
Object {
  "id": 2,
  "meta": Object {
    "cardId": 1,
    "inCardIds": Array [],
  },
  "text": "",
  "userId": "bot",
}
`;

exports[`create a symbol card body 1`] = `
Object {
  "body": Object {
    "id": 3,
    "meta": Array [
      Object {
        "anchorId": 2,
        "createrId": "test-user-1",
        "linenumber": 0,
        "marker": Object {
          "key": "[?]",
          "value": "<BUY> @30",
        },
        "neatReply": true,
        "pollChoices": Array [
          "<BUY>",
        ],
        "stampId": "%3RU",
        "str": " <BUY> @30",
      },
      Object {
        "anchorId": 3,
        "createrId": "test-user-1",
        "linenumber": 2,
        "marker": Object {
          "key": "[+]",
          "value": "111 111",
        },
        "stampId": "%K4q",
        "str": "111 111",
      },
      Object {
        "anchorId": 4,
        "createrId": "test-user-1",
        "linenumber": 3,
        "marker": Object {
          "key": "[+]",
          "value": "222 222",
        },
        "stampId": "%bij",
        "str": "222 222",
      },
    ],
    "prevId": 2,
    "text": "[?] <BUY> @30 %3RU
[+]
111 111 %K4q
222 222 %bij",
    "userId": "test-user-1",
  },
  "bodyId": 3,
  "id": 1,
  "linkUrl": "//$AA",
  "meta": Object {
    "commentId": 2,
    "symbol": "$AA",
  },
  "template": "TICKER",
}
`;

exports[`edit a web card 1`] = `
Object {
  "body": Object {
    "id": 6,
    "meta": Array [
      Object {
        "anchorId": 5,
        "createrId": "test-user-1",
        "linenumber": 1,
        "marker": Object {
          "key": "[*]",
          "value": "2021年會有進一步財政刺激",
        },
        "stampId": "%Jgx",
        "str": "2021年會有進一步財政刺激",
      },
      Object {
        "anchorId": 6,
        "createrId": "test-user-1",
        "linenumber": 2,
        "marker": Object {
          "key": "[*]",
          "value": "歷年有1月效應，1月份呈現反彈，但不認為",
        },
        "stampId": "%PVG",
        "str": "歷年有1月效應，1月份呈現反彈，但不認為",
      },
    ],
    "prevId": 5,
    "text": "[*]
2021年會有進一步財政刺激 %Jgx
歷年有1月效應，1月份呈現反彈，但不認為 %PVG",
    "userId": "test-user-1",
  },
  "bodyId": 6,
  "id": 2,
  "linkUrl": "http://test1.com",
  "meta": Object {
    "commentId": 3,
  },
  "template": "WEBPAGE",
}
`;
